crates/wasi-common/src/dir.rs

```diff
@@ -98,67 +99,72 @@ pub trait WasiDir: Send + Sync {
 }
 
 pub(crate) struct DirEntry {
-    caps: DirCaps,
-    file_caps: FileCaps,
+    caps: RwLock<DirFdStat>,
     preopen_path: Option<PathBuf>, // precondition: PathBuf is valid unicode
     dir: Box<dyn WasiDir>,
 }
 
 impl DirEntry {
     pub fn new(
-        caps: DirCaps,
+        dir_caps: DirCaps,
         file_caps: FileCaps,
         preopen_path: Option<PathBuf>,
         dir: Box<dyn WasiDir>,
     ) -> Self {
         DirEntry {
-            caps,
-            file_caps,
+            caps: RwLock::new(DirFdStat {
+                dir_caps,
+                file_caps,
+            }),
             preopen_path,
             dir,
         }
     }
     pub fn capable_of_dir(&self, caps: DirCaps) -> Result<(), Error> {
-        if self.caps.contains(caps) {
+        if self.caps.read().unwrap().dir_caps.contains(caps) {
             Ok(())
         } else {
-            let missing = caps & !self.caps;
+            let missing = caps & !self.caps.read().unwrap().dir_caps;
             let err = if missing.intersects(DirCaps::READDIR) {
                 Error::not_dir()
             } else {
                 Error::perm()
             };
-            Err(err.context(format!("desired rights {:?}, has {:?}", caps, self.caps)))
+            Err(err.context(format!(
+                "desired rights {:?}, has {:?}",
+                caps,
+                self.caps.read().unwrap().dir_caps
+            )))
         }
     }
     pub fn capable_of_file(&self, caps: FileCaps) -> Result<(), Error> {
-        if self.file_caps.contains(caps) {
+        if self.caps.read().unwrap().file_caps.contains(caps) {
             Ok(())
         } else {
             Err(Error::perm().context(format!(
                 "desired rights {:?}, has {:?}",
-                caps, self.file_caps
+                caps,
+                self.caps.read().unwrap().file_caps
             )))
         }
     }
-    pub fn drop_caps_to(&mut self, caps: DirCaps, file_caps: FileCaps) -> Result<(), Error> {
-        self.capable_of_dir(caps)?;
+    pub fn drop_caps_to(&self, dir_caps: DirCaps, file_caps: FileCaps) -> Result<(), Error> {
+        self.capable_of_dir(dir_caps)?;
         self.capable_of_file(file_caps)?;
-        self.caps = caps;
-        self.file_caps = file_caps;
+        *self.caps.write().unwrap() = DirFdStat {
```

Comment:
Race condition here. Another call to drop_caps_to could have dropped different permissions between the capable_of_{dir,file} call and this write, which would result in capabilities being added back despite the other drop_caps_to returning a success,

Reply:
Thank for spotting this!